src/transaction/atomic_replication_helper.cpp

```diff
@@ -301,160 +354,465 @@ namespace cublog
     return err_code;
   }
 
-  bool
-  atomic_replication_helper::atomic_log_sequence::can_end_sysop_sequence (const LOG_LSA &sysop_parent_lsa) const
+  void
+  atomic_replication_helper::atomic_log_sequence::apply_and_unfix (THREAD_ENTRY *thread_p)
   {
-    if (m_is_sysop && !LSA_ISNULL (&sysop_parent_lsa))
+    if (prm_get_bool_value (PRM_ID_ER_LOG_PTS_ATOMIC_REPL_DEBUG))
       {
-	assert (!m_log_vec.empty ());
-	// if the atomic replication sequence start lsa is higher or equal to the sysop
-	// end parent lsa, then the atomic sequence can be ended (commited & released)
-	return m_start_lsa >= sysop_parent_lsa;
+	dump ("sequence::apply_and_unfix START");
       }
 
-    return false;
-  }
+    // nothing to apply and unfix
+    if (m_log_vec.empty ())
+      {
+	return;
+      }
 
-  bool
-  atomic_replication_helper::atomic_log_sequence::can_end_sysop_sequence () const
-  {
-    if (m_is_sysop)
+    // nothing to apply and unfix; the only existing entry must be the first control log entry
+    if (m_log_vec.size () == 1)
       {
-	assert (!m_log_vec.empty ());
-	return true;
+	assert (m_log_vec[0].is_control ());
+	return;
       }
-    return false;
+
+    atomic_log_entry_vector_type::const_reverse_iterator rbegin_work_log_it = m_log_vec.rbegin ();
+    const atomic_log_entry &rbegin_work_log = *rbegin_work_log_it;
+    if (rbegin_work_log.is_control ())
+      {
+	assert (all_log_entries_are_control ());
+	return;
+      }
+
+    // search backwards for the first non-control log record entry
+    atomic_log_entry_vector_type::const_iterator first_work_log_it = m_log_vec.end ();
+    --first_work_log_it; // last in vector, must be work
+    assert (!first_work_log_it->is_control ());
```

Comment:
Why don't you check here instead of the above check with rbegin_work_log_it (378L-384L)?
```suggestion
        if (first_work_log_it.is_control ())
      {
	assert (all_log_entries_are_control ());
	return;
      }
```

Reply:
Changed as suggested.